lobby-db-dao/src/test/java/org/triplea/lobby/server/db/data/ApiKeyUserDataTest.java

```diff
@@ -0,0 +1,98 @@
+import static org.mockito.Mockito.when;
+
+import java.sql.ResultSet;
+import java.util.Arrays;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class ApiKeyUserDataTest {
+  private static final String ROLE = UserRole.PLAYER;
+  private static final int USER_ID = 55;
+  private static final String NAME = "player-name";
+
+  @Mock private ResultSet resultSet;
+
+  @Test
+  void buildResultMapper() throws Exception {
+    givenResults(USER_ID, ROLE, NAME);
+
+    final ApiKeyUserData result = ApiKeyUserData.buildResultMapper().map(resultSet, null);
+
+    assertThat(result.getUserId(), is(USER_ID));
+    assertThat(result.getRole(), is(UserRole.PLAYER));
+    assertThat(result.getUsername(), is(NAME));
+  }
+
+  private void givenResults(final int userId, final String role, final String name)
+      throws Exception {
+    when(resultSet.getInt(ApiKeyUserData.USER_ID_COLUMN)).thenReturn(userId);
+    when(resultSet.getString(ApiKeyUserData.ROLE_COLUMN)).thenReturn(role);
+    when(resultSet.getString(ApiKeyUserData.USERNAME_COLUMN)).thenReturn(name);
+  }
+
+  @Test
+  void zeroUserIdIsMappedToNull() throws Exception {
+    givenResults(0, UserRole.HOST, null);
+
+    final ApiKeyUserData result = ApiKeyUserData.buildResultMapper().map(resultSet, null);
+
+    assertThat(result.getUserId(), nullValue());
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = {UserRole.PLAYER, UserRole.MODERATOR, UserRole.MODERATOR})
+  void assertInvalidStatesRolesThatMustHaveUserId(final String userRole) throws Exception {
+    givenResults(0, userRole, NAME);
+
+    assertPostconditionFailure();
+  }
+
+  void assertPostconditionFailure() {
+    assertThrows(
+        AssertionError.class, () -> ApiKeyUserData.buildResultMapper().map(resultSet, null));
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = {UserRole.ANONYMOUS, UserRole.HOST})
+  void assertInvalidStatesRolesThatMustNotHaveUserId(final String userRole) throws Exception {
+    givenResults(USER_ID, userRole, NAME);
+
+    assertPostconditionFailure();
+  }
+
+  private static List<List<String>> mustHaveUserName() {
+    return Arrays.asList(
+        Arrays.asList("0", UserRole.ANONYMOUS), //
+        Arrays.asList(String.valueOf(USER_ID), UserRole.PLAYER), //
+        Arrays.asList(String.valueOf(USER_ID), UserRole.MODERATOR), //
+        Arrays.asList(String.valueOf(USER_ID), UserRole.ADMIN));
+  }
+
+  @ParameterizedTest
+  @MethodSource("mustHaveUserName")
+  void assertInvalidStatesRolesThatMustHaveName(final List<String> data) throws Exception {
+    givenResults(Integer.parseInt(data.get(0)), data.get(1), null);
```

Comment:
You can pass multiple arguments using parameterized tests.
I believe the type you'd use instead of a `List` is called `Arguments`, so `mustHaveUsername` would return a `List<Arguments>`
See https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-sources-MethodSource for examples

Reply:
Trying `Arguments` here, access is by object array instead of `List<String>`, lead to more casting:
```
  private static List<Arguments> mustHaveUserName() {
    return Arrays.asList(
        Arguments.of("0", UserRole.ANONYMOUS), //
        Arguments.of(String.valueOf(USER_ID), UserRole.PLAYER), //
        Arguments.of(String.valueOf(USER_ID), UserRole.MODERATOR), //
        Arguments.of(String.valueOf(USER_ID), UserRole.ADMIN));
  }

  @ParameterizedTest
  @MethodSource("mustHaveUserName")
  void assertInvalidStatesRolesThatMustHaveName(final Arguments data) throws Exception {
    givenResults(Integer.parseInt(String.valueOf(data.get()[0])), String.valueOf(data.get()[1]), null);

    assertPostconditionFailure();
  }
```

It looks like from the reference that Arguments is good when we need a stream operation before using the args, ie:
```
public class MyArgumentsProvider implements ArgumentsProvider {

    @Override
    public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
        return Stream.of("apple", "banana").map(Arguments::of);
    }
}
```

The cast to get an integer from `Object` is a bit nasty, as we cast to a string first then to an integer, and then we have another cast to get the second string. `List<String>` is not horrible here, we could do `List<Object>` but then we'll have the same casting issues. Seems a wash, `List<String>` has the least casting.

Reply:
Ah, you misunderstood my point:
When using Arguments the content gets expanded, so you can do this:
```java
  @ParameterizedTest
  @MethodSource("mustHaveUserName")
  void assertInvalidStatesRolesThatMustHaveName(final int id, final String role) throws Exception {
    givenResults(id, role, null);

    assertPostconditionFailure();
  }
```

Reply:
Of course this makes the assumption that you the types match the types you passed to the arguments object

Reply:
Ah, that was good to learn about, I did not get that from the documentation. It is much cleaner with `Arguments`, no casting at all, updated: 2fcae9823

Reply:
AFAIK it works with a simple 2 dimensional array or other container structures as well